File: android/src/main/java/com/facebook/flipper/plugins/databases/ObjectMapper.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.plugins.databases;

import android.text.TextUtils;
import com.facebook.flipper.core.FlipperArray;
import com.facebook.flipper.core.FlipperArray.Builder;
import com.facebook.flipper.core.FlipperObject;
import com.facebook.flipper.plugins.databases.DatabaseDriver.DatabaseExecuteSqlResponse;
import com.facebook.flipper.plugins.databases.DatabaseDriver.DatabaseGetTableDataResponse;
import com.facebook.flipper.plugins.databases.DatabaseDriver.DatabaseGetTableInfoResponse;
import com.facebook.flipper.plugins.databases.DatabaseDriver.DatabaseGetTableStructureResponse;
import com.facebook.flipper.plugins.databases.DatabasesManager.DatabaseDescriptorHolder;
import com.facebook.flipper.plugins.databases.DatabasesManager.ExecuteSqlRequest;
import com.facebook.flipper.plugins.databases.DatabasesManager.GetTableDataRequest;
import com.facebook.flipper.plugins.databases.DatabasesManager.GetTableInfoRequest;
import com.facebook.flipper.plugins.databases.DatabasesManager.GetTableStructureRequest;
import java.io.UnsupportedEncodingException;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class ObjectMapper {

  private static final int MAX_BLOB_LENGTH = 100 * 1024;
  private static final String UNKNOWN_BLOB_LABEL_FORMAT = "{%d-byte %s blob}";

  public static FlipperArray databaseListToFlipperArray(
      Collection<DatabaseDescriptorHolder> databaseDescriptorHolderList) {
    FlipperArray.Builder builder = new FlipperArray.Builder();

    for (DatabaseDescriptorHolder databaseDescriptorHolder : databaseDescriptorHolderList) {

      List<String> tableNameList =
          databaseDescriptorHolder.databaseDriver.getTableNames(
              databaseDescriptorHolder.databaseDescriptor);
      Collections.sort(tableNameList);
      FlipperArray.Builder tableBuilder = new Builder();
      for (String tablename : tableNameList) {
        tableBuilder.put(tablename);
      }

      builder.put(
          new FlipperObject.Builder()
              .put("id", databaseDescriptorHolder.id)
              .put("name", databaseDescriptorHolder.databaseDescriptor.name())
              .put("tables", tableBuilder.build())
              .build());
    }

    return builder.build();
  }

  public static GetTableDataRequest flipperObjectToGetTableDataRequest(FlipperObject params) {
    int databaseId = params.getInt("databaseId");
    String table = params.getString("table");
    String order = params.getString("order");
    boolean reverse = params.getBoolean("reverse");
    int start = params.getInt("start");
    int count = params.getInt("count");
    if (databaseId <= 0 || TextUtils.isEmpty(table)) {
      return null;
    }
    return new GetTableDataRequest(databaseId, table, order, reverse, start, count);
  }

  public static GetTableStructureRequest flipperObjectToGetTableStructureRequest(
      FlipperObject params) {
    int databaseId = params.getInt("databaseId");
    String table = params.getString("table");
    if (databaseId <= 0 || TextUtils.isEmpty(table)) {
      return null;
    }
    return new GetTableStructureRequest(databaseId, table);
  }

  public static GetTableInfoRequest flipperObjectToGetTableInfoRequest(FlipperObject params) {
    int databaseId = params.getInt("databaseId");
    String table = params.getString("table");
    if (databaseId <= 0 || TextUtils.isEmpty(table)) {
      return null;
    }
    return new GetTableInfoRequest(databaseId, table);
  }

  public static ExecuteSqlRequest flipperObjectToExecuteSqlRequest(FlipperObject params) {
    int databaseId = params.getInt("databaseId");
    String value = params.getString("value");
    if (databaseId <= 0 || TextUtils.isEmpty(value)) {
      return null;
    }
    return new ExecuteSqlRequest(databaseId, value);
  }

  public static FlipperObject databaseGetTableDataReponseToFlipperObject(
      DatabaseGetTableDataResponse databaseGetTableDataResponse) {

    FlipperArray.Builder columnBuilder = new FlipperArray.Builder();
    for (String columnName : databaseGetTableDataResponse.columns) {
      columnBuilder.put(columnName);
    }

    FlipperArray.Builder rowBuilder = new FlipperArray.Builder();
    for (List<Object> row : databaseGetTableDataResponse.values) {
      FlipperArray.Builder valueBuilder = new FlipperArray.Builder();
      for (Object item : row) {
        valueBuilder.put(objectAndTypeToFlipperObject(item));
      }
      rowBuilder.put(valueBuilder.build());
    }

    return new FlipperObject.Builder()
        .put("columns", columnBuilder.build())
        .put("values", rowBuilder.build())
        .put("start", databaseGetTableDataResponse.start)
        .put("count", databaseGetTableDataResponse.count)
        .put("total", databaseGetTableDataResponse.total)
        .build();
  }

  public static FlipperObject databaseGetTableStructureResponseToFlipperObject(
      DatabaseGetTableStructureResponse databaseGetTableStructureResponse) {

    FlipperArray.Builder structureColumnBuilder = new FlipperArray.Builder();
    for (String columnName : databaseGetTableStructureResponse.structureColumns) {
      structureColumnBuilder.put(columnName);
    }

    FlipperArray.Builder indexesColumnBuilder = new FlipperArray.Builder();
    for (String columnName : databaseGetTableStructureResponse.indexesColumns) {
      indexesColumnBuilder.put(columnName);
    }

    FlipperArray.Builder structureValuesBuilder = new FlipperArray.Builder();
    for (List<Object> row : databaseGetTableStructureResponse.structureValues) {
      FlipperArray.Builder valueBuilder = new FlipperArray.Builder();
      for (Object item : row) {
        valueBuilder.put(objectAndTypeToFlipperObject(item));
      }
      structureValuesBuilder.put(valueBuilder.build());
    }

    FlipperArray.Builder indexesValuesBuilder = new FlipperArray.Builder();
    for (List<Object> row : databaseGetTableStructureResponse.indexesValues) {
      FlipperArray.Builder valueBuilder = new FlipperArray.Builder();
      for (Object item : row) {
        valueBuilder.put(objectAndTypeToFlipperObject(item));
      }
      indexesValuesBuilder.put(valueBuilder.build());
    }

    return new FlipperObject.Builder()
        .put("structureColumns", structureColumnBuilder.build())
        .put("structureValues", structureValuesBuilder.build())
        .put("indexesColumns", indexesColumnBuilder.build())
        .put("indexesValues", indexesValuesBuilder.build())
        .build();
  }

  public static FlipperObject databaseGetTableInfoResponseToFlipperObject(
      DatabaseGetTableInfoResponse databaseGetTableInfoResponse) {

    return new FlipperObject.Builder()
        .put("definition", databaseGetTableInfoResponse.definition)
        .build();
  }

  public static FlipperObject databaseExecuteSqlResponseToFlipperObject(
      DatabaseExecuteSqlResponse databaseExecuteSqlResponse) {

    FlipperArray.Builder columnBuilder = new FlipperArray.Builder();
    if (databaseExecuteSqlResponse.columns != null) {
      for (String columnName : databaseExecuteSqlResponse.columns) {
        columnBuilder.put(columnName);
      }
    }

    FlipperArray.Builder rowBuilder = new FlipperArray.Builder();
    if (databaseExecuteSqlResponse.values != null) {
      for (List<Object> row : databaseExecuteSqlResponse.values) {
        FlipperArray.Builder valueBuilder = new FlipperArray.Builder();
        for (Object item : row) {
          valueBuilder.put(objectAndTypeToFlipperObject(item));
        }
        rowBuilder.put(valueBuilder.build());
      }
    }

    return new FlipperObject.Builder()
        .put("type", databaseExecuteSqlResponse.type)
        .put("columns", columnBuilder.build())
        .put("values", rowBuilder.build())
        .put("insertedId", databaseExecuteSqlResponse.insertedId)
        .put("affectedCount", databaseExecuteSqlResponse.affectedCount)
        .build();
  }

  private static FlipperObject objectAndTypeToFlipperObject(Object object) {
    if (object == null) {
      return new FlipperObject.Builder().put("type", "null").build();
    } else if (object instanceof Long) {
      return new FlipperObject.Builder().put("type", "integer").put("value", object).build();
    } else if (object instanceof Double) {
      return new FlipperObject.Builder().put("type", "float").put("value", object).build();
    } else if (object instanceof String) {
      return new FlipperObject.Builder().put("type", "string").put("value", object).build();
    } else if (object instanceof byte[]) {
      return new FlipperObject.Builder()
          .put("type", "blob")
          .put("value", blobToString((byte[]) object))
          .build();
    } else if (object instanceof Boolean) {
      return new FlipperObject.Builder().put("type", "boolean").put("value", object).build();
    } else {
      throw new IllegalArgumentException("type of Object is invalid");
    }
  }

  private static String blobToString(byte[] blob) {
    if (blob.length <= MAX_BLOB_LENGTH) {
      if (fastIsAscii(blob)) {
        try {
          return new String(blob, "US-ASCII");
        } catch (UnsupportedEncodingException e) {
          // Fall through...
        }
      } else {
        // try UTF-8
        try {
          return new String(blob, "UTF-8");
        } catch (UnsupportedEncodingException e) {
          // Fall through...
        }
      }
      return String.format(Locale.US, UNKNOWN_BLOB_LABEL_FORMAT, blob.length, "binary");
    }
    return String.format(Locale.US, UNKNOWN_BLOB_LABEL_FORMAT, blob.length, "large");
  }

  private static boolean fastIsAscii(byte[] blob) {
    for (byte b : blob) {
      if ((b & ~0x7f) != 0) {
        return false;
      }
    }
    return true;
  }

  public static FlipperObject toErrorFlipperObject(int code, String message) {
    return new FlipperObject.Builder().put("code", code).put("message", message).build();
  }
}

File: iOS/Plugins/FlipperKitDatabasesPlugin/FlipperKitDatabasesPlugin/ObjectMapper.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "ObjectMapper.h"
#include <Foundation/Foundation.h>
#import "DatabaseDescriptorHolder.h"
#import "DatabaseExecuteSql.h"
#import "DatabaseGetTableData.h"
#import "DatabaseGetTableInfo.h"
#import "DatabaseGetTableStructure.h"

@implementation ObjectMapper

static const int MAX_BLOB_LENGTH = 100 * 1024;
static NSString* const UNKNOWN_BLOB_LABEL_FORMAT = @"{%d-byte %@ blob}";

+ (NSMutableArray*)databaseListToFlipperArray:
    (NSMutableSet<DatabaseDescriptorHolder*>*)databaseDescriptorHolderSet {
  NSMutableArray* result = [NSMutableArray new];

  for (DatabaseDescriptorHolder* holder in databaseDescriptorHolderSet) {
    NSArray<NSString*>* tables =
        [holder.databaseDriver getTableNames:holder.databaseDescriptor];
    NSArray<NSString*>* sortedTableNames =
        [tables sortedArrayUsingSelector:@selector(compare:)];
    NSString* idString = [NSString stringWithFormat:@"%ld", holder.identifier];

    NSDictionary* databaseInfo = @{
      @"id" : idString,
      @"name" : holder.databaseDescriptor.name,
      @"tables" : sortedTableNames
    };
    [result addObject:databaseInfo];
  }

  return result;
}

+ (NSDictionary*)databaseGetTableDataResponseToDictionary:
    (DatabaseGetTableDataResponse*)response {
  NSMutableArray* rows = [NSMutableArray array];
  for (NSArray* row in response.values) {
    NSMutableArray* rowValues = [NSMutableArray array];
    for (id item in row) {
      [rowValues addObject:[self objectAndTypeToFlipperObject:item]];
    }
    [rows addObject:rowValues];
  }

  return @{
    @"columns" : response.columns,
    @"values" : rows,
    @"start" : @(response.start),
    @"count" : @(response.count),
    @"total" : @(response.total)
  };
}

+ (NSDictionary*)errorWithCode:(NSInteger)code message:(NSString*)message {
  return @{@"code" : @(code), @"message" : message};
}

+ (NSDictionary*)databaseGetTableStructureResponseToDictionary:
    (DatabaseGetTableStructureResponse*)response {
  NSMutableArray* structureValues = [NSMutableArray array];
  for (NSArray* row in response.structureValues) {
    NSMutableArray* rowValues = [NSMutableArray array];
    for (id item in row) {
      [rowValues addObject:[self objectAndTypeToFlipperObject:item]];
    }
    [structureValues addObject:rowValues];
  }

  NSMutableArray* indexesValues = [NSMutableArray array];
  for (NSArray* row in response.indexesValues) {
    NSMutableArray* rowValues = [NSMutableArray array];
    for (id item in row) {
      [rowValues addObject:[self objectAndTypeToFlipperObject:item]];
    }
    [indexesValues addObject:rowValues];
  }

  return @{
    @"structureColumns" : response.structureColumns,
    @"structureValues" : structureValues,
    @"indexesColumns" : response.indexesColumns,
    @"indexesValues" : indexesValues
  };
}

+ (NSDictionary*)databaseGetTableInfoResponseToDictionary:
    (DatabaseGetTableInfoResponse*)response {
  return @{
    @"definition" : response.definition,
  };
}

+ (NSDictionary*)databaseExecuteSqlResponseToDictionary:
    (DatabaseExecuteSqlResponse*)response {
  NSMutableArray* rows = [NSMutableArray array];
  if (response.values) {
    for (NSArray* row in response.values) {
      NSMutableArray* rowValues = [NSMutableArray array];
      for (id item in row) {
        [rowValues addObject:[self objectAndTypeToFlipperObject:item]];
      }
      [rows addObject:rowValues];
    }
  }

  NSMutableDictionary* result = [NSMutableDictionary dictionaryWithDictionary:@{
    @"type" : response.type,
    @"columns" : response.columns,
    @"values" : rows,
    @"affectedCount" : @(response.affectedCount)
  }];

  if (response.insertedId) {
    result[@"insertedId"] = response.insertedId;
  }

  return result;
}

+ (NSDictionary*)objectAndTypeToFlipperObject:(id)object {
  if (!object || [object isKindOfClass:[NSNull class]]) {
    return @{@"type" : @"null"};
  } else if ([object isKindOfClass:[NSNumber class]]) {
    NSNumber* number = (NSNumber*)object;
    NSString* type = [NSString stringWithUTF8String:[number objCType]];

    if ([type isEqualToString:@"i"]) {
      return @{@"type" : @"integer", @"value" : number};
    } else if ([type isEqualToString:@"f"] || [type isEqualToString:@"d"]) {
      return @{@"type" : @"float", @"value" : number};
    } else if ([type isEqualToString:@"B"]) {
      return @{@"type" : @"boolean", @"value" : number};
    } else {
      return @{@"type" : @"integer", @"value" : @([number integerValue])};
    }

    return @{@"type" : @"integer", @"value" : object};
  } else if ([object isKindOfClass:[NSDecimalNumber class]]) {
    return @{@"type" : @"float", @"value" : object};
  } else if ([object isKindOfClass:[NSString class]]) {
    return @{@"type" : @"string", @"value" : object};
  } else if ([object isKindOfClass:[NSData class]]) {
    NSString* blobString = [self blobToString:(NSData*)object];
    return @{@"type" : @"blob", @"value" : blobString};
  } else if ([object isKindOfClass:[NSDictionary class]]) {
    // Usualy the dictionary is a Json blob, and we can parse it as string.
    NSError* error;
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:object
                                                       options:0
                                                         error:&error];
    if (!jsonData) {
      NSString* reason = [NSString
          stringWithFormat:@"NSDictionary is not in a json format: %@",
                           [error localizedDescription]];
      @throw [NSException exceptionWithName:@"InvalidArgumentException"
                                     reason:reason
                                   userInfo:nil];
    }

    NSString* jsonString = [[NSString alloc] initWithData:jsonData
                                                 encoding:NSUTF8StringEncoding];
    return @{@"type" : @"blob", @"value" : jsonString};

  } else if ([object isKindOfClass:[NSValue class]]) {
    return @{@"type" : @"boolean", @"value" : object};
  } else {
    @throw [NSException exceptionWithName:@"InvalidArgumentException"
                                   reason:@"type of Object is invalid"
                                 userInfo:nil];
  }
}

+ (NSString*)blobToString:(NSData*)data {
  const uint8_t* bytes = data.bytes;
  uint length = data.length;

  if (length <= MAX_BLOB_LENGTH) {
    if ([self fastIsAscii:bytes length:length]) {
      NSStringEncoding encoding = NSASCIIStringEncoding;
      return [[NSString alloc] initWithBytesNoCopy:(void*)bytes
                                            length:length
                                          encoding:encoding
                                      freeWhenDone:NO];
    } else {
      // try UTF-8
      NSStringEncoding encoding = NSUTF8StringEncoding;
      return [[NSString alloc] initWithBytesNoCopy:(void*)bytes
                                            length:length
                                          encoding:encoding
                                      freeWhenDone:NO];
    }
  }
  return
      [NSString stringWithFormat:UNKNOWN_BLOB_LABEL_FORMAT, length, @"binary"];
}

+ (BOOL)fastIsAscii:(const uint8_t*)bytes length:(NSUInteger)length {
  for (int i = 0; i < length; i++) {
    uint8_t b = bytes[i];
    if ((b & ~0x7f) != 0) {
      return NO;
    }
  }
  return YES;
}

@end
